### PR TITLE
955 clear out uses of addAction in tests

### LINKED ISSUE #955: Clear out uses of addAction in tests/
**What Needs to be Done?**

We want to largely deprecate casual usage of `theTerm()->addAction`, in favor of higher-level routines instead. The dominant replacement for a lot of uses should probably be `vt::runInEpochCollective()`, with the associated action simply sequenced afterward inline.

```
$ git grep -c addAction tests/
tests/unit/active/test_active_send.cc:1
tests/unit/active/test_pending_send.cc:1
tests/unit/collection/test_checkpoint.extended.cc:1
tests/unit/collection/test_destroy.cc:1
tests/unit/collection/test_insert.extended.cc:6
tests/unit/collectives/test_mpi_collective.cc:3
tests/unit/group/test_group.cc:4
tests/unit/location/test_hops.extended.cc:1
tests/unit/location/test_location_common.h:1
tests/unit/memory/test_memory_active.cc:2
tests/unit/memory/test_memory_lifetime.cc:14
tests/unit/objgroup/test_objgroup.cc:3
tests/unit/pipe/test_callback_bcast.cc:6
tests/unit/pipe/test_callback_bcast_collection.extended.cc:3
tests/unit/pipe/test_callback_func.cc:1
tests/unit/pipe/test_callback_func_ctx.cc:1
tests/unit/pipe/test_callback_send.cc:6
tests/unit/pipe/test_callback_send_collection.extended.cc:3
tests/unit/rdma/test_rdma_collection_handle.extended.cc:1
tests/unit/sequencer/test_sequencer.extended.cc:4
tests/unit/sequencer/test_sequencer_extensive.extended.cc:1
tests/unit/sequencer/test_sequencer_for.extended.cc:1
tests/unit/sequencer/test_sequencer_nested.extended.cc:1
tests/unit/sequencer/test_sequencer_parallel.extended.cc:2
tests/unit/sequencer/test_sequencer_vrt.extended.cc:3
tests/unit/termination/test_termination_action_callable.extended.cc:2
tests/unit/termination/test_termination_action_common.impl.h:2
tests/unit/termination/test_termination_reset.cc:2
```

There are 77 such calls in total. The ones in `tests/unit/termination` may be appropriate to leave as is.

Jonathan and I have decided that we'd like to split up this effort among the new-comers, to provide a bit of an exercise, and spread out some experience with the commit, PR, and CI workflow.

https://github.com/DARMA-tasking/vt/pull/953/files#diff-3ab9ba90e26bdf35497656cb95df5b39 gives an illustration of the desired kind of change

I'll post comments with sub-assignments. Please each submit a separate PR, and assign it to me @PhilMiller for review.

**Is your feature request related to a problem? Please describe.**

**Describe potential solution outcome**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by cz4rs**
Fixes #955

### CODE REVIEW

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/memory/test_memory_lifetime.cc  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5979299&bid=19606446)

**Comment by cz4rs** [comment]
there's still a couple of files to be cleared, the rest is landing on Monday

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/memory/test_memory_lifetime.cc  -4
+ tests/unit/memory/test_memory_active.cc  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5979299&bid=19606446)

File: tests/unit/memory/test_memory_active.cc [hunk 1]
```diff
@@ -83,28 +83,22 @@ TYPED_TEST_P(TestMemoryActive, test_memory_remote_send) {
 
   std::vector<MsgSharedPtr<MsgType>> msgs;
 
-  if (run_test && my_node == 0) {
```
Comment (lifflander) [inline tests/unit/memory/test_memory_active.cc:86]: This is an example where we can't use the new API. This should be the same as before.

File: tests/unit/memory/test_memory_active.cc [hunk 2]
```diff
@@ -114,28 +108,23 @@ TYPED_TEST_P(TestMemoryActive, test_memory_remote_broadcast) {
 
   std::vector<MsgSharedPtr<MsgType>> msgs;
 
-  if (num_nodes > 1) {
```
Comment (lifflander) [inline tests/unit/memory/test_memory_active.cc:117]: Same here

File: tests/unit/memory/test_memory_lifetime.cc [hunk 3]
```diff
@@ -219,22 +209,17 @@ TEST_F(TestMemoryLifetime, test_active_send_normal_lifetime_2) {
 TEST_F(TestMemoryLifetime, test_active_bcast_normal_lifetime_1) {
   auto const& num_nodes = theContext()->getNumNodes();
 
-  if (num_nodes > 1) {
-    for (int i = 0; i < num_msgs_sent; i++) {
-      auto msg = makeMessage<NormalTestMsg>();
-      theMsg()->broadcastMsg<NormalTestMsg, normalHan>(msg.get());
-
-      theTerm()->addAction([msg]{
```
Comment (lifflander) [inline tests/unit/memory/test_memory_lifetime.cc:227]: Same here. We need the action and we must have finalization of event.

File: tests/unit/memory/test_memory_lifetime.cc [hunk 4]
```diff
@@ -243,22 +228,17 @@ TEST_F(TestMemoryLifetime, test_active_bcast_normal_lifetime_1) {
 TEST_F(TestMemoryLifetime, test_active_bcast_normal_lifetime_2) {
   auto const& num_nodes = theContext()->getNumNodes();
 
-  if (num_nodes > 1) {
-    for (int i = 0; i < num_msgs_sent; i++) {
-      auto msg = makeMessage<NormalTestMsg>();
-      theMsg()->broadcastMsg<NormalTestMsg, normalHan>(msg.get());
-
-      theTerm()->addAction([msg]{
```
Comment (lifflander) [inline tests/unit/memory/test_memory_lifetime.cc:251]: Same here.

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/memory/test_memory_active.cc  -2
+ tests/unit/memory/test_memory_lifetime.cc  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5979299&bid=19606446)

**Comment by cz4rs** [comment]
files cleared:

- [ ] tests/unit/group/test_group.cc:4
- [ ] tests/unit/location/test_hops.extended.cc:1
- [ ] tests/unit/location/test_location_common.h:1
- [ ] tests/unit/memory/test_memory_active.cc:2
- [ ] tests/unit/memory/test_memory_lifetime.cc:14
- [ ] tests/unit/objgroup/test_objgroup.cc:3
- [ ] tests/unit/pipe/test_callback_bcast.cc:6
- [ ] tests/unit/pipe/test_callback_bcast_collection.extended.cc:3

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/memory/test_memory_lifetime.cc  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5979299&bid=19606446)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/memory/test_memory_lifetime.cc  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5979299&bid=19606446)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/memory/test_memory_lifetime.cc  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5979299&bid=19606446)

**Comment by cz4rs** [comment]
documentation improvements:

image
- remove `copydoc` for `vt::sched::Scheduler::scheduler` (which results in random "Turn the scheduler")
note: it would be better to improve documentation of the method instead
- remove duplication by omitting `\brief` (use `copydetails` instead of `copydoc`)

File: tests/unit/memory/test_memory_active.cc (hunk 1 above)
Comment (cz4rs) [reply tests/unit/memory/test_memory_active.cc]: reverted

File: tests/unit/memory/test_memory_active.cc (hunk 2 above)
Comment (cz4rs) [reply tests/unit/memory/test_memory_active.cc]: reverted

File: tests/unit/memory/test_memory_lifetime.cc (hunk 3 above)
Comment (cz4rs) [reply tests/unit/memory/test_memory_lifetime.cc]: reverted

File: tests/unit/memory/test_memory_lifetime.cc (hunk 4 above)
Comment (cz4rs) [reply tests/unit/memory/test_memory_lifetime.cc]: reverted

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/memory/test_memory_lifetime.cc  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5979299&bid=19606446)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/memory/test_memory_lifetime.cc  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5979299&bid=19606446)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/memory/test_memory_lifetime.cc  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5979299&bid=19606446)

**Comment by cz4rs** [comment]
@lifflander 
> This looks great now, I think it's ready to merge once CI passes

Yep, I think it's good to go :)

Side note: you might have seen that already, TimeTrigger tests are still failing from time to time (`clang-8, macosx, mpich` build)
```
The following tests FAILED:
	354 - vt:TestTimeTrigger.test_time_trigger_2_proc_4 (Failed)
```
plus there's one (new?) warning generated from that test:
```
 [295/480] Building CXX object tests/CMakeFiles/test_time_trigger_extended.dir/unit/timetrigger/test_time_trigger.extended.cc.o
../../tests/unit/timetrigger/test_time_trigger.extended.cc:121:41: warning: lambda capture 'cur_time' is not used [-Wunused-lambda-capture]
      trigger_period[i], [&triggered,i,&cur_time]{
                                      ~~^~~~~~~~
```

**Comment by lifflander** [comment]
> @lifflander
> 
> > This looks great now, I think it's ready to merge once CI passes
> 
> Yep, I think it's good to go :)
> 
> Side note: you might have seen that already, TimeTrigger tests are still failing from time to time (`clang-8, macosx, mpich` build)
> 
> ```
> The following tests FAILED:
> 	354 - vt:TestTimeTrigger.test_time_trigger_2_proc_4 (Failed)
> ```
> 
> plus there's one (new?) warning generated from that test:
> 
> ```
>  [295/480] Building CXX object tests/CMakeFiles/test_time_trigger_extended.dir/unit/timetrigger/test_time_trigger.extended.cc.o
> ../../tests/unit/timetrigger/test_time_trigger.extended.cc:121:41: warning: lambda capture 'cur_time' is not used [-Wunused-lambda-capture]
>       trigger_period[i], [&triggered,i,&cur_time]{
>                                       ~~^~~~~~~~
> ```

Yes, I saw that it is still failing. Planning to fix

**Comment by lifflander** [comment]
@cz4rs Ok, can you rebase and then we can merge?

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/memory/test_memory_lifetime.cc  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5979299&bid=19606446)

**Comment by cz4rs** [comment]
> @cz4rs Ok, can you rebase and then we can merge?

rebased with the latest changes from `develop`

File: tests/unit/objgroup/test_objgroup.cc [hunk 5]
```diff
@@ -172,65 +172,66 @@ TEST_F(TestObjGroup, test_proxy_update) {
...
 
-  auto obj = proxy.get();
-  vt_debug_print(
-    objgroup, node,
-    "obj->recv:{} before term\n", obj->recv_
-  );
+  runInEpochCollective([&]{
+    // self-send a message and then broadcast
+    proxy[my_node].send<MyMsg, &MyObjA::handler>();
+    proxy.broadcast<MyMsg, &MyObjA::handler>();
```
Comment (PhilMiller) [inline tests/unit/objgroup/test_objgroup.cc:184]: @lifflander I think this is a scalability problem for cases where vt is configured and compiled on a host with large numbers of cores - the message count will scale with the number of cores, which can make it take an arbitrarily long time on a very wide node

File: tests/unit/location/test_location_common.h [hunk 6]
```diff
@@ -151,18 +151,13 @@ void verifyCacheConsistency(
     // perform the checks only at the end of the epoch
     // to ensure that all entity messages have been
     // correctly delivered before.
```
Comment (PhilMiller) [inline tests/unit/location/test_location_common.h:153]: @lifflander I think this change may have fundamentally shifted the meaning of this test. Please double check

Comment (lifflander) [reply tests/unit/location/test_location_common.h:153]: This diff is really confusing to read. Had to go back and look.

This test is not testing the same thing at all. It's now checking the state before. I should have caught this.

Comment (cz4rs) [reply tests/unit/location/test_location_common.h:153]: I will revert and (try to) handle this correctly in #989

Comment (cz4rs) [reply tests/unit/location/test_location_common.h:153]: efe6ed595cc28a9485270f50e1526955ee4a8e0d added in #989

**Pull request merged**